### PR TITLE
feat: support custom OIDC token endpoint URL

### PR DESCRIPTION
**Pull request by mikesouza**
## Description

This PR addresses the inconsistency in how the JS SDK constructs the OIDC token endpoint URL (see https://github.com/openfga/sdk-generator/issues/238), so that it is inline with how the Go and other SDKs work.

- Supports custom token endpoint paths specified by the `apiTokenIssuer`.
- Configuration of the `apiTokenIssuer` now accepts a full URL that can optionally include the scheme. If a scheme is not specified, it defaults to `https://`.
- Added and updated unit tests to ensure backwards compatibility.

## References

* closes https://github.com/openfga/js-sdk/issues/141
* replaces/closes #271 

NOTE: This PR #271 seems to be outdated and the author doesn't seem to be working on it, so I decided to submit my own PR.

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [X] The correct base branch is being used, if not main
- [X] I have added tests to validate that the change in functionality is working as expected

## Summary by CodeRabbit

* **Improvements**
  * Enhanced credential handling to support flexible API token issuer formats, including automatic endpoint path normalization and improved validation for various URL schemes.

* **Tests**
  * Added comprehensive test coverage for token endpoint resolution, validating multiple issuer formats, URL schemes, and endpoint path handling.

### CODE REVIEW

File: credentials/credentials.ts [hunk 1]
```diff
@@ -138,13 +140,35 @@ export class Credentials {
...
+    if (url) {
+      if (url.pathname === "" || url.pathname === "/") {
+        url.pathname = DEFAULT_TOKEN_ENDPOINT_PATH;
+      }
+
+      return url.toString();
+    }
+
+    return "";
+  }
```
[the remark this answers is not in the document]
Comment (mikesouza) [reply credentials/credentials.ts]: Checking the scheme is not necessary here because the parsed URL will be falsey/null if a scheme is not specified.

File: credentials/credentials.ts [hunk 2]
```diff
@@ -138,13 +140,38 @@ export class Credentials {
...
+   * Defaults to https:// scheme if none provided and appends the default
+   * token endpoint path when the issuer has no path or only a root path.
+   * 
+   * @param apiTokenIssuer
+   * @return string The constructed token endpoint URL, or empty string if invalid
+   */
+  private buildApiTokenUrl(apiTokenIssuer: string): string {
+    let url = URL.parse(apiTokenIssuer);
+    if (!url && !apiTokenIssuer.startsWith("https://") && !apiTokenIssuer.startsWith("http://")) {
+      url = URL.parse(`https://${apiTokenIssuer}`);
```
Comment (ttrzeng) [inline credentials/credentials.ts]: buildApiTokenUrl is using URL.parse, but this file doesn’t import the Node url module, and global URL doesn’t have .parse. 

Either:
`import the legacy url module (import * as url from "url"; const parsed = url.parse(...))` or
just use the standard new URL(...) approach

File: credentials/credentials.ts [hunk 3]
```diff
@@ -93,9 +95,9 @@ export class Credentials {
       assertParamExists("Credentials", "config.apiAudience", authConfig.config?.apiAudience);
       assertParamExists("Credentials", "config.clientSecret or config.clientAssertionSigningKey", (authConfig.config as ClientSecretConfig).clientSecret || (authConfig.config as PrivateKeyJWTConfig).clientAssertionSigningKey);
 
-      if (!isWellFormedUriString(`https://${authConfig.config?.apiTokenIssuer}`)) {
+      if (!isWellFormedUriString(this.buildApiTokenUrl(authConfig.config?.apiTokenIssuer))) {
         throw new FgaValidationError(
-          `Configuration.apiTokenIssuer does not form a valid URI (https://${authConfig.config?.apiTokenIssuer})`);
+          `Configuration.apiTokenIssuer does not form a valid URI (${authConfig.config?.apiTokenIssuer})`);
```
[the remark this answers is not in the document]
Comment (mikesouza) [reply credentials/credentials.ts]: What if I just pass to `isWellFormedUriString(this.normalizeApiTokenIssuer(apiTokenIssuer))` instead?

File: credentials/credentials.ts [hunk 4]
```diff
@@ -93,9 +96,9 @@ export class Credentials {
       assertParamExists("Credentials", "config.apiAudience", authConfig.config?.apiAudience);
       assertParamExists("Credentials", "config.clientSecret or config.clientAssertionSigningKey", (authConfig.config as ClientSecretConfig).clientSecret || (authConfig.config as PrivateKeyJWTConfig).clientAssertionSigningKey);
 
-      if (!isWellFormedUriString(`https://${authConfig.config?.apiTokenIssuer}`)) {
+      if (!isWellFormedUriString(this.normalizeApiTokenIssuer(authConfig.config?.apiTokenIssuer))) {
         throw new FgaValidationError(
-          `Configuration.apiTokenIssuer does not form a valid URI (https://${authConfig.config?.apiTokenIssuer})`);
+          `Configuration.apiTokenIssuer does not form a valid URI (${this.normalizeApiTokenIssuer(authConfig.config?.apiTokenIssuer)})`);
```
Comment (mikesouza) [inline credentials/credentials.ts]: I just call it again here instead of storing to a variable due to `Unexpected lexical declaration in case block` linting rule.

File: tests/credentials.test.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  OPENFGA_CLIENT_ID,
+  OPENFGA_CLIENT_SECRET,
+} from "./helpers/default-config";
+
+nock.disableNetConnect();
+
+describe("Credentials", () => {
+  const mockTelemetryConfig: TelemetryConfiguration = new TelemetryConfiguration({});
+
+  describe("Refreshing access token", () => {
```
Comment (daniel-jonathan) [inline tests/credentials.test.ts]: These tests verify the token endpoint URL construction, but don't verify the JWT aud claim for PrivateKeyJWT flow. Consider adding tests that verify the audience matches the normalized issuer for different schemes (http://, https://, no scheme).

File: credentials/credentials.ts [hunk 6]
```diff
@@ -138,13 +141,49 @@ export class Credentials {
...
+   * @param apiTokenIssuer
+   * @return string The constructed token endpoint URL if valid, otherwise throws an error
+   * @throws {FgaValidationError} If the API token issuer URL is invalid
+   */
+  private buildApiTokenUrl(apiTokenIssuer: string): string {
+    const normalizedApiTokenIssuer = this.normalizeApiTokenIssuer(apiTokenIssuer);
+
+    try {
+      const url = new URL(normalizedApiTokenIssuer);
+      if (url.pathname === "" || url.pathname === "/") {
```
Comment (daniel-jonathan) [inline credentials/credentials.ts]: Consider adding test coverage for edge cases:
- Multiple trailing slashes: `https://issuer.fga.example///`
- Path with only slashes: `https://issuer.fga.example//`
- Empty string `apiTokenIssuer`
- Whitespace-only issuer

**Potential issue:** The current logic only checks for `""` or `"/"`, but `url.pathname` could be `"//"` or other variants which wouldn't trigger the default path assignment.

**Suggested fix:**
```javascript
if (!url.pathname || url.pathname.match(/^\/+$/)) {
  // pathname is empty or only contains slashes
  url.pathname = `/${DEFAULT_TOKEN_ENDPOINT_PATH}`;
}

Comment (mikesouza) [reply credentials/credentials.ts]: I'm not opposed to covering more edge cases, but just to be clear, this would intentionally diverge from the behavior of the [Go SDK](https://github.com/openfga/go-sdk/blob/main/credentials/credentials.go#L146-L160) and maybe the other SDKs. I can make this change, but just want to check if consistency in behavior in SDKs is preferred.

**Review by daniel-jonathan — APPROVED** [review]

**Comment by mikesouza** [comment]
@daniel-jonathan Thanks for the re-review!

Hmmm strange, the unit tests all pass on my local machine, but it looks like the [test failures in CI](https://github.com/openfga/js-sdk/actions/runs/19574293291/job/56065637811) are related to the 15 tests that use `nock`.  🤔  

image

**Review by ttrzeng — DISMISSED** [review]

**Review by SoulPancake — DISMISSED** [review]
LGTM

**Comment by mikesouza** [comment]
@SoulPancake Thank you, I see you were trying earlier to debug and fix my tests to figure out why they are failing in CI but pass locally. Any luck on finding the issue?

I'm baffled at what's different compared to the other test files with tests that use nock. The only thing I could see different is I'm using table tests (`test.each`), which shouldn't make a difference, but at this point I'm out of guesses. 

I pushed this [commit](https://github.com/openfga/js-sdk/pull/285/commits/685f2af2ddabfc04498f9d6fd22399362b71de36) to not use table tests to see if it passes in CI.
